platform/darwin/src/NSExpression+MGLAdditions.mm

```diff
@@ -61,4 +61,30 @@ @implementation NSExpression (MGLAdditions)
     return { };
 }
 
+- (mbgl::FeatureIdentifier)mgl_featureIdentifier
+{
+    id value = self.constantValue;
+    mbgl::Value mbglValue = [self mgl_filterValue];
+    
+    if ([value isKindOfClass:NSString.class]) {
+        return mbglValue.get<std::string>();
+    } else if ([value isKindOfClass:NSNumber.class]) {
+        NSNumber *number = (NSNumber *)value;
+        if ((strcmp([number objCType], @encode(char)) == 0) ||
+            (strcmp([number objCType], @encode(BOOL)) == 0)) {
+            return mbglValue.get<bool>();
+        } else if ( strcmp([number objCType], @encode(double)) == 0 ||
+                    strcmp([number objCType], @encode(float)) == 0) {
+            return mbglValue.get<double>();
+        } else if ([number compare:@(0)] == NSOrderedDescending ||
+                   [number compare:@(0)] == NSOrderedSame) {
+            return mbglValue.get<uint64_t>();
+        } else if ([number compare:@(0)] == NSOrderedAscending) {
+            return mbglValue.get<int64_t>();
+        }
+    }
```

Comment:
I thought about this and decided to not throw an exception when I added it yesterday. 

The reason is that the first thing `mgl_featureIdentifier` does is call `mgl_convertedValueWithValue` which will throw an exception if there is an unknown type. `mgl_featureIdentifier` exists only to be able to get the correct raw value from `mbgl::Value` returned by `mgl_convertedValueWithValue` and the only reason we are doing any of this here is to share the conversion logic that happens to currently live in the `NSExpression` category.

I think you could make the case that we should catch an exception thrown by `mgl_convertedValueWithValue` and re-raise it with a different message (instead of the `Can’t convert %s:%@ to mbgl::Value`) that a developer will see if they send an invalid value for the `id` / `identifier`. However, I was thinking that it might make sense to wait and refactor this if we introduce a category method on NSObject as you've suggested in conversation. It is a bit weird at the moment though.